lib/reversetunnel/seek/state.go

```diff
@@ -0,0 +1,211 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package seek
+
+import (
+	"time"
+)
+
+// seekState represents the state of a seeker.
+type seekState string
+
+const (
+	// stateNil is the zero value of seekState
+	stateNil seekState = ""
+	// stateSeeking indicates that we want to connect to this proxy
+	stateSeeking seekState = "seeking"
+	// stateClaimed indicates that an agent has successfully claimed
+	// responsibility for this proxy
+	stateClaimed seekState = "claimed"
+	// stateBackoff indicates that this proxy was claimed but that the
+	// agent responsible for it lost the connection prematurely
+	stateBackoff seekState = "backoff"
+)
+
+// seekEvent represents an asynchronous event which may change
+// the state of a seeker.
+type seekEvent string
+
+const (
+	// eventNil is the zero value of seekEvent
+	eventNil seekEvent = ""
+	// eventRefresh indicates that a proxy has been indirectly observed (e.g. via gossip).
+	eventRefresh seekEvent = "refresh"
+	// eventAcquire indicates that an agent has connected to this proxy and would like to
+	// take responsibility for it.
+	eventAcquire seekEvent = "acquire"
+	// eventRelease indicates that the agent responsible for this proxy has lost its
+	// connection to it.
+	eventRelease seekEvent = "release"
+)
+
+// seeker manages the state associated with a proxy.
+type seeker struct {
+	state   seekState
+	at      time.Time
+	backOff uint64
+}
+
+// transit attempts a state transition.  If transit returns true, then
+// a state-transition did occur.
+func (s *seeker) transit(t time.Time, e seekEvent, c *Config) (ok bool) {
+	// eventNil cannot trigger a state transition
+	if e == eventNil {
+		return false
+	}
+	// map stateNil to stateSeeking since stateSeeking is the
+	// default initial state.
+	if s.state == stateNil {
+		s.state = stateSeeking
+	}
+	switch s.state {
+	case stateSeeking:
+		// stateSeeking can either transition to stateClaimed, or
+		// be "refreshed" in order to prevent expiration.
+		switch e {
+		case eventRefresh:
+			if t.After(s.at) {
+				s.at = t
+				return true
+			} else {
+				return false
+			}
+		case eventAcquire:
+			s.state = stateClaimed
+			s.at = t
+			return true
+		case eventRelease:
+			return false
+		default:
+			panic("unreachable")
+		}
+	case stateClaimed:
+		// stateClaimed can either transition into stateSeeking or
+		// stateBackoff depending on whether the claim failed
+		// immediately, or after some period of normal operation.
+		switch e {
+		case eventRefresh, eventAcquire:
+			return false
+		case eventRelease:
+			if s.shouldBackoff(t, c.BackoffThreshold) {
```

Comment:
add a little more rationale comment here - why this is necessary (to protect from unhealthy proxies scenarios)